common/persistence/Pinot/pinotVisibilityStore.go

```diff
@@ -0,0 +1,1111 @@
+			DomainUUID:    request.DomainUUID,
+			Domain:        request.Domain,
+			EarliestTime:  time.Time{},
+			LatestTime:    time.Time{},
+			NextPageToken: request.NextPageToken,
+			PageSize:      request.PageSize,
+		},
+	}
+
+	return v.pinotClient.Search(req)
+}
+
+func (v *pinotVisibilityStore) CountWorkflowExecutions(ctx context.Context, request *p.CountWorkflowExecutionsRequest) (*p.CountWorkflowExecutionsResponse, error) {
+	query := v.getCountWorkflowExecutionsQuery(v.pinotClient.GetTableName(), request)
+
+	resp, err := v.pinotClient.CountByQuery(query)
+	if err != nil {
+		return nil, &types.InternalServiceError{
+			Message: fmt.Sprintf("CountClosedWorkflowExecutions failed, %v", err),
+		}
+	}
+
+	return &p.CountWorkflowExecutionsResponse{
+		Count: resp,
+	}, nil
+}
+
+func (v *pinotVisibilityStore) checkProducer() {
+	if v.producer == nil {
+		// must be bug, check history setup
+		panic("message producer is nil")
+	}
+}
+
+// a new function to create visibility message for deletion
+// don't use the other function and provide some nil values because it may cause nil pointer exceptions
+func createDeleteVisibilityMessage(domainID string,
+	wid,
+	rid string,
+	isDeleted bool,
+) (*indexer.PinotMessage, error) {
+	m := make(map[string]interface{})
+	m[DocID] = wid + "-" + rid
+	m[DomainID] = domainID
+	m[WorkflowID] = wid
+	m[RunID] = rid
+	m[IsDeleted] = isDeleted
+	serializedMsg, err := json.Marshal(m)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := &indexer.PinotMessage{
+		WorkflowID: common.StringPtr(wid),
+		Payload:    serializedMsg,
+	}
+	return msg, nil
+}
+
+func createVisibilityMessage(
+	// common parameters
+	domainID string,
+	wid,
+	rid string,
+	workflowTypeName string,
+	taskList string,
+	startTimeUnixNano int64,
+	executionTimeUnixNano int64,
+	taskID int64,
+	memo []byte,
+	encoding common.EncodingType,
+	isCron bool,
+	numClusters int16,
+	// specific to certain status
+	closeTimeUnixNano int64, // close execution
+	closeStatus workflow.WorkflowExecutionCloseStatus, // close execution
+	historyLength int64, // close execution
+	updateTimeUnixNano int64, // update execution,
+	shardID int64,
+	rawSearchAttributes map[string][]byte,
+	isDeleted bool,
+) (*indexer.PinotMessage, error) {
+	m := make(map[string]interface{})
+	//loop through all input parameters
+	m[DocID] = wid + "-" + rid
+	m[DomainID] = domainID
+	m[WorkflowID] = wid
+	m[RunID] = rid
+	m[WorkflowType] = workflowTypeName
+	m[TaskList] = taskList
+	m[StartTime] = startTimeUnixNano
+	m[ExecutionTime] = executionTimeUnixNano
+	m[IsCron] = isCron
+	m[NumClusters] = numClusters
+	m[CloseTime] = closeTimeUnixNano
+	m[CloseStatus] = int(closeStatus)
+	m[HistoryLength] = historyLength
+	m[UpdateTime] = updateTimeUnixNano
+	m[ShardID] = shardID
+	m[IsDeleted] = isDeleted
+
+	SearchAttributes := make(map[string]interface{})
+	var err error
+	for key, value := range rawSearchAttributes {
+		value, err = isTimeStruct(value)
```

Comment:
i'm not familiar with pinot. Does it not support time.Time type?

Reply:
From the [official doc](https://docs.pinot.apache.org/basics/components/table/schema#date-and-time-fields), it is not supported

> Since Pinot doesn't have a dedicated DATETIME datatype support, you need to input time in either STRING, LONG, or INT format. However, Pinot needs to convert the date into an understandable format such as epoch timestamp to do operations.